cmd/transcriber/call/tracks.go

```diff
@@ -0,0 +1,341 @@
+				log.Printf("fixing audio timestamp by %d", gap)
+			}
+		}
+
+		prevArrivalTime = time.Now()
+		prevRTPTimestamp = pkt.Timestamp
+
+		if err := oggWriter.WriteRTP(pkt, gap); err != nil {
+			log.Printf("failed to write RTP packet: %s", err)
+		}
+	}
+}
+
+// handleClose will kick off post-processing of saved voice tracks.
+func (t *Transcriber) handleClose(_ any) error {
+	log.Printf("handleClose")
+
+	defer close(t.doneCh)
+
+	t.liveTracksWg.Wait()
+	close(t.trackCtxs)
+
+	log.Printf("live tracks processing done, starting post processing")
+	start := time.Now()
+
+	var samplesDur time.Duration
+	var tr transcribe.Transcription
+	for ctx := range t.trackCtxs {
+		log.Printf("post processing track %s", ctx.trackID)
+
+		trackTr, dur, err := t.transcribeTrack(ctx)
+		if err != nil {
+			log.Printf("failed to transcribe track %q: %s", ctx.trackID, err)
+			continue
+		}
+
+		samplesDur += dur
+
+		tr = append(tr, trackTr)
+	}
+
+	if len(tr) == 0 {
+		return fmt.Errorf("nothing to do, empty transcription")
+	}
+
+	dur := time.Since(start)
+	log.Printf("transcription process completed for all tracks: transcribed %v of audio in %v, %0.2fx",
+		samplesDur, dur, samplesDur.Seconds()/dur.Seconds())
+
+	f, err := os.OpenFile(filepath.Join(getDataDir(), fmt.Sprintf("%s-%s.vtt",
+		t.cfg.CallID, time.Now().UTC().Format("2006-01-02-15_04_05"))), os.O_RDWR|os.O_CREATE, 0600)
+	if err != nil {
+		return fmt.Errorf("failed to open output file: %w", err)
+	}
+	defer f.Close()
+
+	if err := tr.WebVTT(f); err != nil {
+		return fmt.Errorf("failed to write WebVTT file: %w", err)
+	}
+
+	if err := t.publishTranscription(f); err != nil {
+		return fmt.Errorf("failed to publish transcription: %w", err)
+	}
+
+	log.Printf("transcription published successfully")
+
+	return nil
+}
+
+// trackTimedSamples is used to account for potential gaps in
+// voice tracks due to mute/unmute sequences. Each spoken segment
+// will have a relative time offset (startTS).
+type trackTimedSamples struct {
+	pcm     []float32
+	startTS int64
+}
+
+// decodeAudio reads a track OGG file and decodes its audio into raw PCM samples
+// for later processing.
+func (ctx trackContext) decodeAudio() ([]trackTimedSamples, error) {
+	trackFile, err := os.Open(ctx.filename)
+	defer trackFile.Close()
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to open track file: %w", err)
+	}
+
+	oggReader, _, err := ogg.NewReaderWith(trackFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new ogg reader: %w", err)
+	}
+
+	opusDec, err := opus.NewDecoder(trackOutAudioRate, trackAudioChannels)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create opus decoder: %w", err)
+	}
+	defer func() {
+		if err := opusDec.Destroy(); err != nil {
+			log.Printf("failed to destroy decoder: %s", err)
+		}
+	}()
+
+	log.Printf("decoding track %s", ctx.trackID)
+
+	pcmBuf := make([]float32, trackOutFrameSize)
```

Comment:
Is this where we decide the framesize for the samples (the chunk that whisper uses to transcribe)? Does that mean this is the entire context that whisper has to deal with, or does it keep the past samples as context internally? (not sure how it works, as you can see)
Is there an optimal framesize for quality of transcription output?

Reply:
> Is this where we decide the framesize for the samples (the chunk that whisper uses to transcribe)?

Nope, that is just the space that a single opus audio packet will need for decoding. All the decoded raw PCM samples are then appended below:

https://github.com/mattermost/calls-transcriber/blob/254603fd5a374f0c4676ce9adc2d2b70a799f87a/cmd/transcriber/call/tracks.go#L283

> or does it keep the past samples as context internally?

Whisper has indeed an option to keep past transcription context to improve the accuracy and we are making use of that:

https://github.com/mattermost/calls-transcriber/blob/254603fd5a374f0c4676ce9adc2d2b70a799f87a/cmd/transcriber/apis/whisper.cpp/context.go#L71

We could also tweak it further using the [`n_max_text_ctx`](https://github.com/ggerganov/whisper.cpp/blob/a4bb2df36aeb4e6cfb0c1ca9fbcf749ef39cc852/whisper.h#L355) param but so far going with the default [`16384`](https://github.com/ggerganov/whisper.cpp/blob/21e8c67a4fb6bfc5f1af2a6c00db97b28a0f15b2/whisper.cpp#L3456C12-L3456C26).

> Is there an optimal framesize for quality of transcription output?

Generally the more you can pass in one go the better. And that's what we are doing when possible, considering the constrains of having to deal with tracks separately and with potential gaps in between.

Reply:
Great, thank you for the help, it's clearer now!